tests/base/quadrature_point_data_04.cc

```diff
@@ -0,0 +1,247 @@
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+// This test checks whether ContinuousQuadratureDatatransfer allows
+// each cell to have different number of quadrature point data.
+
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/base/quadrature_point_data.h>
+
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_values.h>
+
+#include <deal.II/grid/grid_generator.h>
+
+#include "../tests.h"
+
+using namespace dealii;
+
+using MaterialBase = TransferableQuadraturePointData;
+using QuadratureStorage =
+  CellDataStorage<typename Triangulation<2>::cell_iterator, MaterialBase>;
+
+// history structs for each material
+struct Mat0 : MaterialBase
+{
+  double x;
+  virtual unsigned int
+  number_of_values() const final
+  {
+    return 1;
+  }
+  virtual void
+  pack_values(std::vector<double> &values) const final
+  {
+    AssertDimension(values.size(), number_of_values());
+    values[0] = x;
+  }
+  virtual void
+  unpack_values(const std::vector<double> &values) final
+  {
+    AssertDimension(values.size(), number_of_values());
+    x = values[0];
+  }
+};
+
+struct Mat1 : MaterialBase
+{
+  Point<2> pt;
+  virtual unsigned int
+  number_of_values() const final
+  {
+    return 2;
+  }
+  virtual void
+  pack_values(std::vector<double> &values) const final
+  {
+    AssertDimension(values.size(), number_of_values());
+    std::copy(pt.begin_raw(), pt.end_raw(), values.begin());
+  }
+  virtual void
+  unpack_values(const std::vector<double> &values) final
+  {
+    AssertDimension(values.size(), number_of_values());
+    std::copy(values.cbegin(), values.cend(), pt.begin_raw());
+  }
+};
+
+// This material has zero-size data
+struct Mat2 : MaterialBase
+{
+  virtual unsigned int
+  number_of_values() const final
+  {
+    return 0;
+  }
+  virtual void
+  pack_values(std::vector<double> &values) const final
+  {
+    AssertDimension(values.size(), number_of_values());
+  }
+  virtual void
+  unpack_values(const std::vector<double> &values) final
+  {
+    AssertDimension(values.size(), number_of_values());
+  }
+};
+
+constexpr unsigned int n_quadrature_per_dim = 3;
+constexpr unsigned int n_data_points_per_cell =
+  n_quadrature_per_dim * n_quadrature_per_dim;
+
+void
+initialize_data(const Triangulation<2> &tria, QuadratureStorage &storage)
+{
+  deallog << "Initializing quadrature cell data" << std::endl;
+  for (auto cell : tria.active_cell_iterators())
+    if (cell->is_locally_owned())
+      {
+        // It does not modify the data on already initialized cell data
+        if (cell->material_id() == 0)
+          storage.template initialize<Mat0>(cell, n_data_points_per_cell);
+        else if (cell->material_id() == 1)
+          storage.template initialize<Mat1>(cell, n_data_points_per_cell);
+        else if (cell->material_id() == 2)
+          storage.template initialize<Mat2>(cell, n_data_points_per_cell);
+      }
```

Comment:
Move the comment `// material_id 3 does not have an associated data structure` here.